prometheus_juju_exporter/collector.py

```diff
@@ -140,6 +140,40 @@ def _get_machine_type(self, machine: Dict) -> MachineType:
 
         return MachineType.METAL
 
+    def _get_machine_identifier(self, machine: Dict) -> str:
+        """Try to find a valid identifier for the machine.
+
+        The hostname field is only supported for juju controller versions
+        2.8.10 and upwards. For the remaining lower versions, this field
+        exists however is None or "None". This method tries to get it from
+        hostname if it can, else it will try to get it from the "instance-id"
+        field, which seems to exist almost always.
+
+        :param dict machine: status information for a machine
+        :return str: a valid identifier string for the machine if
+            applicable, else the string literal "None".
+        """
+        self.logger.debug("Trying to find an identifier for machine: %s", str(machine))
+        machine_id = "None"
+        for field in ["hostname", "instance-id"]:
+            candidate = machine.get(field, None)
+            self.logger.debug(
+                "Candidate machine id:[%s] in field:[%s]", candidate, field
+            )
+            if candidate not in [None, "None", "pending"]:
+                machine_id = candidate
+                self.logger.debug(
+                    "Selecting machine id:[%s] in field:[%s]",
+                    machine_id,
+                    field,
+                )
+                break
+        if machine_id == "None":
+            self.logger.error(
+                "Failed to find a machine identifier for machine: %s", str(machine)
```

Comment:
I'm thinking we should handle `pending` in a different way. With your current code, if a machine is pending, we would have `machine_id = "None"` and log it as an error. A warning (or even info?) would probably be better in this scenario.

Reply:
updated